eyeshade/lib/transaction.js

```diff
@@ -269,30 +273,57 @@ async function insertFromVoting (runtime, client, voteDoc, surveyorCreatedAt) {
         return
       }
 
-      const query = `
-      insert into transactions ( id, created_at, description, transaction_type, document_id, from_account, from_account_type, to_account, to_account_type, amount, channel )
-      VALUES ( $1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11 )
-      `
-      await runtime.postgres.query(query, [
+      return [
         // surveyorId and channel pair should be unique
         uuidv5(voteDoc.surveyorId + normalizedChannel, 'be90c1a8-20a3-4f32-be29-ed3329ca8630'),
         surveyorCreatedAt,
         `votes from ${voteDoc.surveyorId}`,
         'contribution',
         voteDoc.surveyorId,
-        runtime.config.wallet.settlementAddress.BAT,
+        settlementAddress,
         'uphold',
         normalizedChannel,
         'channel',
         amount.plus(fees).toString(),
         normalizedChannel
-      ], client)
+      ]
     }
   } else {
     throw new Error('Missing amount field')
   }
 }
 
+async function insertManyFromVoting (atOneTime, runtime, client, docs, surveyorCreatedAt) {
+  for (let i = 0; i < docs.length; i += atOneTime) {
+    const mapped = docs.slice(i, i + atOneTime).map((doc) =>
+      insertFromVotingArguments(
+        runtime.config.wallet.settlementAddress.BAT,
+        doc,
+        surveyorCreatedAt
+      )
+    )
+    const query = `
+    insert into transactions ( id, created_at, description, transaction_type, document_id, from_account, from_account_type, to_account, to_account_type, amount, channel )
+    VALUES %L`
+    await runtime.postgres.query(format(query, mapped), null, client, {
+      text: query,
+      length: mapped.length
+    })
+  }
+}
```

Comment:
You either need to put an assertion at the beginning of the function that `docs.length` is divisible by batch size, or you need special handling to insert what's left after the loop. Pulling that out into an closure is how [we do it in stats](https://github.com/brave/stats/blob/b8746ea289adc3dcc9cff2d9d3a01f68068c1a28/src/services/common-aggregation.service.js#L151-L178).

Reply:
resolved in DM.